### ISSUE TITLE
Authenticator Username form validation does not match Cognito's requirements

### ISSUE THREAD

**Issue by Jordan-Nelson** [aws-amplify/amplify-flutter#2188]
### Description

Cognito uses the regex `[\p{L}\p{M}\p{S}\p{N}\p{P}]` for validate usernames. The Authenticator allows for any non-empty string.

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [ ] Auth
- [X] Authenticator
- [ ] DataStore
- [ ] Storage

### Steps to Reproduce

1. Set up an app with username/password based auth.
2. Try to sign up with a username that has a white space in the middle
3. **Actual behavior:** Request is sent to cognito, cognito error message is displayed.
4. **Expected behavior:** Form validation prevents the request from being sent to Cognito and displays a form validation message.

### Platforms

- [X] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Environment

```bash
N/A
```

### Dependencies

```bash
N/A
```

### Device

N/A

### OS

N/A

### Deployment Method

Amplify CLI

### Amplify Config

N/A

**Comment by Lorenzohidalgo** [comment]
@Jordan-Nelson & @salihgueler I'll take a shot at this issue 😉

**Comment by Lorenzohidalgo** [comment]
Hey @Jordan-Nelson, could you confirm that the provided regex expression is correct?

Using `[\p{L}\p{M}\p{S}\p{N}\p{P}]+` will allow users to create usernames with whitespaces. On the contrary, using `^[\p{L}\p{M}\p{S}\p{N}\p{P}\S*]+$` won't allow usernames with spaces, which would make more sense under my point of view.

The current regex for the email address `^\S+@\S+$` also allows formats like `hello@jordan` as valid input. An expression like `^[^\s@]+@[^\s@]+\.[^\s@]+$` would be a little more accurate. Should I update that as well?

**Comment by Jordan-Nelson** [comment]
@Lorenzohidalgo - Yes, `^[\p{L}\p{M}\p{S}\p{N}\p{P}\S*]+$` would be appropriate. Thanks for calling that out.

> The current regex for the email address ^\S+@\S+$ also allows formats like hello@jordan as valid input. An expression like ^[^\s@]+@[^\s@]+\.[^\s@]+$ would be a little more accurate. Should I update that as well?

Cognito actually allows `hello@jordan` as an email. It does not allow `hello` as an email though. My understanding is that the regex below is the agreed upon standard in the community for what is considered a valid email address. Below are also a few sources for that. 

Regex:
```
/^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/
```

Sources:
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#basic_validation
- https://www.w3.org/TR/2017/REC-html52-20171214/sec-forms.html#valid-e-mail-address
- https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address

The regex used in the authenticator is mostly inline with that. `^\S+@\S+$` allows for some additional characters than what is listed above. I would have to do some digging/testing to see which of the two Cognito follows.

While I am not aware of a valid use case for omitting the extension (.com, .net, etc.), staying in line with the community / Cognito makes sense.

tl;dr - I think we can leave the regex for email unless you are finding it less restrictive than Cognito, in which case I can do some digging and find out exactly what Cognito is using.

**Comment by Lorenzohidalgo** [comment]
Perfect @Jordan-Nelson , I was unaware that this was the standard, will use the updated username regex expression on the PR and leave the email one as-is.

**Comment by neetable-barath** [comment]
We are still facing this issue in flutter amplify_authenticator: [1.2.1](https://pub.dev/packages/amplify_authenticator/versions/1.2.1) latest version

**Comment by Jordan-Nelson** [comment]
@neetable-barath could you provide some more details? If you have some examples of usernames that would help.

**Comment by dgagnon** [comment]
a note regarding domain validation: what about IDN ?  Most of these regexp will completely exclude them.  Does cognito supports using IDN ?

ref:  https://en.wikipedia.org/wiki/Internationalized_domain_name
